module/zfs/spa.c

```diff
@@ -8121,8 +8132,12 @@ spa_async_remove(spa_t *spa, vdev_t *vd)
 
 		vdev_state_dirty(vd->vdev_top);
 
-		/* Tell userspace that the vdev is gone. */
-		zfs_post_remove(spa, vd);
+		if (vd->vdev_zed_event) {
+			vd->vdev_zed_event = B_FALSE;
+		} else {
+			/* Tell userspace that the vdev is gone. */
+			zfs_post_remove(spa, vd);
+		}
```

Comment:
Does thins cause some event loop otherwise, or what?  I am thinking what if some other process rather than one that sent the removal request want to receive this notification?

Reply:
Exactly, this is used to prevent the event loop if this is sent from zed. If ```vd->vdev_zed_event``` is set it means that vdev is removed already and zed knows about that.

Reply:
For the reason I specified I'd prefer the loop to be broken on side of the process requesting removal.  But that is just a feeling, I'll leave it to somebody with better understanding of the whole picture.

Reply:
I think you are right. I will deal with this from the ZED context.